### PR TITLE
🧹 Reduce duplication in snippet testing

### PR DESCRIPTION
**Pull request by rix0rrr**
The snippet testing code had some duplication. In this code, we separate out how to find the interesting snippets in the YAML files. For every snippet, we remember where in a YAML file we found it (by keeping track of the keys we used to descend into dictionaries and the indexes we used to descend into lists).

That way, the "restore to English" code can become generic: since every snippet knows where in the YAML structure it comes from, we can automatically find the same snippet in the English language file and copy it over.

**How to test**

If the tests pass, everything works!

### CODE REVIEW

File: tests/test_snippets/snippet_tester.py [hunk 1]
```diff
@@ -235,19 +212,112 @@ def do_snippet(self, snippet, yaml_locator=None):
...
 
-    def restore_snippet_to_english(self, snippet, yaml_locator):
+    def restore_snippet_to_english(self, snippet):
         # English file is always 'en.yaml' in the same dir
         en_file = path.join(path.dirname(snippet.filename), 'en.yaml')
 
         # Read English yaml file
         original_yaml = YamlFile.for_file(en_file)
-        original_loc = yaml_locator(snippet, original_yaml)
+        original_loc = locate_snippet_in_yaml(original_yaml, snippet, yaml_locator)
```
Comment (boryanagoncharenko) [inline tests/test_snippets/snippet_tester.py:221]: The `yaml_locator` does not need to be supplied any more

File: tests/test_snippets/snippet_tester.py [hunk 2]
```diff
@@ -9,18 +9,17 @@
 - Snippets can hold on to the information where in the YAML file they were
```
Comment (boryanagoncharenko) [inline tests/test_snippets/snippet_tester.py:9]: This could be scratched now, right :)

**Review by boryanagoncharenko — CHANGES_REQUESTED** [review]

File: tests/test_snippets/snippet_tester.py [hunk 3]
```diff
@@ -30,165 +29,146 @@
...
+            if tag.name == 'pre' and tag.contents and tag.contents[0].contents]
 
 
 def filter_snippets(snippets, level=None, lang=None):
     if (lang or level) and os.getenv('CI'):
         raise RuntimeError('Whoops, it looks like you left a snippet filter in!')
 
     if lang:
-        snippets = [(name, snippet) for (name, snippet) in snippets if snippet.language[:2] == lang]
+        snippets = [s for s in snippets if x.language[:len(lang)] == lang]
```
[the remark this answers is not in the document]
Comment (rix0rrr) [reply tests/test_snippets/snippet_tester.py:156]: Oops!

[the remark this answers is not in the document]
Comment (rix0rrr) [reply tests/test_snippets/snippet_tester.py:156]: Tiens. I thought `pylint` would catch errors like these.

**Comment by rix0rrr** [comment]
Good catches @boryanagoncharenko, thanks! I'm a bit surprised that the static analysis tools we run didn't catch these, since they're pretty obvious. 

I've confirmed that `pylint` doesn't catch them (grr). `pylance` could potentially do it, but I didn't find out how to run it.

We should probably find a static checker to catch trivial errors?

File: tests/test_snippets/snippet_tester.py (hunk 2 above)
Comment (rix0rrr) [reply tests/test_snippets/snippet_tester.py:9]: Ha. Of course!

**Comment by rix0rrr** [comment]
I've re-added `pylint` to the build checks, and confirmed that `pylint` would have caught those errors.

We used to have it, not sure why we removed it?

**Comment by rix0rrr** [comment]
Shit, pylint was removed in this PR: https://github.com/hedyorg/hedy/pull/5134

@jpelay, do you remember why?

**Comment by rix0rrr** [comment]
Doesn't seem to be easy to re-enable. My struggles will continue in here: https://github.com/hedyorg/hedy/pull/5377

In the mean time, this one should be good to go.

**Review by boryanagoncharenko — APPROVED** [review]

**Pull request merged**